### PR TITLE
Feature/saa 1083 update waiting list endpoint

### PR DESCRIPTION
**Pull request by matt-crowson-moj**
The is the first draft/cut of a new endpoint to allow for updating waiting list applications.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/request/WaitingListApplicationUpdateRequest.kt [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  val status: WaitingListStatus? = null,
+) {
+  @AssertTrue(message = "Requested by cannot be empty or blank")
+  private fun isRequestedBy() = requestedBy == null || requestedBy.isNotBlank()
+
+  @AssertTrue(message = "Application date cannot be in the future")
+  private fun isApplicationDate() = applicationDate == null || applicationDate <= LocalDate.now()
+
+  @AssertTrue(message = "Only PENDING, APPROVED or DECLINED are allowed for status")
+  private fun isStatus() =
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/request/WaitingListApplicationUpdateRequest.kt:56]: Can they move any of these statuses to any other of these? No workflow? So can go from DECLINED to APPROVED?

Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/request/WaitingListApplicationUpdateRequest.kt:56]: As far as I am aware no.  Only the 3 statues via the workflow in the screens at the mo.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/WaitingListService.kt [hunk 2]
```diff
@@ -69,6 +73,15 @@ class WaitingListService(
       .also { log.info("Added ${request.status} waiting list application for prisoner ${request.prisonerNumber} to activity ${schedule.description}") }
   }
 
+  private fun WaitingListApplicationRequest.failIfNotAllowableStatus() {
```
[the remark this answers is not in the document]
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/WaitingListService.kt:76]: Added before the model change.  It does no harm.  I'd be inclined to leave it in.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/WaitingListService.kt [hunk 3]
```diff
@@ -125,4 +138,77 @@ class WaitingListService(
...
+      .failIfNotUpdatable()
+      .apply {
+        updateApplicationDate(request, updatedBy)
+        updateRequestedBy(request, updatedBy)
+        updateComments(request, updatedBy)
+        updateStatus(request, updatedBy)
+      }.toModel()
+
+  private fun WaitingList.failIfNotUpdatable() = also {
+    require(isStatus(WaitingListStatus.APPROVED, WaitingListStatus.PENDING, WaitingListStatus.DECLINED)) {
```
[the remark this answers is not in the document]
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/WaitingListService.kt:156]: Added before the model change. It does no harm. I'd be inclined to leave it in.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/WaitingListService.kt [hunk 4]
```diff
@@ -125,4 +138,77 @@ class WaitingListService(
...
+    request.status?.takeUnless { it == status }?.let { updatedStatus ->
+      require(
+        listOf(
+          WaitingListStatus.APPROVED,
+          WaitingListStatus.PENDING,
+          WaitingListStatus.DECLINED,
+        ).contains(updatedStatus),
+      ) {
+        "Only PENDING, APPROVED or DECLINED are allowed for the status change"
+      }
```
[the remark this answers is not in the document]
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/WaitingListService.kt:203]: This one can probably go.

**Review by timharrison-moj — APPROVED** [review]
LGTM

**Pull request merged**